### PR TITLE
[aon_timer,dv] Improvements on AON timer smoke test

### PR DESCRIPTION
**Pull request by ctopal**
This commit includes some DV enhancements of AON timer module.
Which includes:
- A proper initialization and shutdown tasks.
- A core interface to see the timer counts.
- A wait for interrupt so that we would only exit when any
meaningful thing happens.

Signed-off-by: Canberk Topal <ctopal@lowrisc.org>

### CODE REVIEW

File: hw/ip/aon_timer/dv/env/aon_timer_core_if.sv [hunk 1]
```diff
@@ -2,47 +2,12 @@
...
-  input logic [31:0]               wdog_bark_thold_wr_data_i,
-  input logic                      wdog_bite_thold_reg_wr_i,
-  input logic [31:0]               wdog_bite_thold_wr_data_i,
-  input logic                      wdog_count_reg_wr_i,
-  input logic [31:0]               wdog_count_wr_data_i,
-
-  input logic                      wkup_intr_o,
-  input logic                      wdog_intr_o,
-  input logic                      wdog_reset_req_o
+  input logic                      wkup_count_reg_wr_o,
```
Comment (sriyerg) [inline hw/ip/aon_timer/dv/env/aon_timer_core_if.sv:8]: How are these signals used?

File: hw/ip/aon_timer/dv/env/aon_timer_env.sv [hunk 2]
```diff
@@ -32,6 +32,10 @@ class aon_timer_env extends cip_base_env #(
         get(this, "", "sleep_vif", cfg.sleep_vif)) begin
       `uvm_fatal(`gfn, "failed to get sleep_vif from uvm_config_db")
     end
+    if (!uvm_config_db#(virtual aon_timer_core_if)::get(this, "", "aon_timer_core_vif",
+                                                         cfg.aon_timer_core_vif)) begin
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/aon_timer_env.sv:36]: Nit: Extra space of indentation

File: hw/ip/aon_timer/dv/env/aon_timer_env_cfg.sv [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
 class aon_timer_env_cfg extends cip_base_env_cfg #(.RAL_T(aon_timer_reg_block));
 
+  virtual aon_timer_core_if aon_timer_core_vif;
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/aon_timer_env_cfg.sv:7]: Not a problem, but I'd probably just call this `core_vif`: no need for the aon_timer "namespacing"

File: hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv [hunk 4]
```diff
@@ -65,14 +65,11 @@ class aon_timer_scoreboard extends cip_base_scoreboard #(
...
-      "intr_enable": begin
-        // FIXME
-      end
       "intr_test": begin
         // FIXME
       end
       default: begin
-        `uvm_fatal(`gfn, $sformatf("invalid csr: %0s", csr.get_full_name()))
+          // No other special behaviour for writes
       end
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv:73]: If you're dropping the "case for each CSR" approach, I'd suggest dropping `intr_test` here too.

File: hw/ip/aon_timer/dv/env/seq_lib/aon_timer_base_vseq.sv [hunk 5]
```diff
@@ -11,7 +11,7 @@ class aon_timer_base_vseq extends cip_base_vseq #(
   `uvm_object_utils(aon_timer_base_vseq)
 
   // various knobs to enable certain routines
-  rand bit do_aon_timer_init;
+  bit do_aon_timer_init = 0;
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/seq_lib/aon_timer_base_vseq.sv:14]: Is this bit actually used any more?

File: hw/ip/aon_timer/dv/env/seq_lib/aon_timer_base_vseq.sv [hunk 6]
```diff
@@ -31,17 +31,51 @@ class aon_timer_base_vseq extends cip_base_vseq #(
...
+    // Clear the interrupts
+    csr_utils_pkg::csr_wr(ral.intr_state, 2'b11);
+
+    // Zero out the COUNT registers
+    csr_utils_pkg::csr_wr(ral.wkup_count, 32'h0000_0000);
+    csr_utils_pkg::csr_wr(ral.wdog_count, 32'h0000_0000);
+
+    `uvm_info(`gfn, "Disabling AON Timer. Writing 0 to WKUP_CTRL and WDOG_CTRL", UVM_LOW)
+    csr_utils_pkg::csr_wr(ral.wkup_ctrl.enable, 1'b0);
+    csr_utils_pkg::csr_wr(ral.wdog_ctrl.enable, 1'b0);
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/seq_lib/aon_timer_base_vseq.sv:52]: This is fine, but I'm not sure I'd actually bother! Do we have any situations where we use this logic?

File: hw/ip/aon_timer/dv/env/seq_lib/aon_timer_base_vseq.sv [hunk 7]
```diff
@@ -31,17 +31,51 @@ class aon_timer_base_vseq extends cip_base_vseq #(
...
   endtask
 
   // setup basic aon_timer features
   virtual task aon_timer_init();
-    `uvm_info(`gfn, "Initializating aon timer, nothing to do at the moment", UVM_MEDIUM)
+    uint wkup_thold;
+    uint wdog_bark_thold;
+    uint wdog_bite_thold;
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/seq_lib/aon_timer_base_vseq.sv:59]: I'd consider turning these into fields of the sequence and adding constraints there, which means that you don't have to randomise them explicitly.

File: hw/ip/aon_timer/dv/env/seq_lib/aon_timer_base_vseq.sv [hunk 8]
```diff
@@ -66,4 +100,14 @@ class aon_timer_base_vseq extends cip_base_vseq #(
...
   endtask
+
+  task wait_for_interrupt();
+    if (cfg.aon_clk_rst_vif.rst_n &&
+        !cfg.aon_intr_vif.pins[0] &&
+        !cfg.aon_intr_vif.pins[1]) begin
+      @(negedge cfg.aon_clk_rst_vif.rst_n or
+        posedge cfg.aon_intr_vif.pins[0] or
+        posedge cfg.aon_intr_vif.pins[1]);
+    end
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/seq_lib/aon_timer_base_vseq.sv:111]: I think I might be lazier:
```suggestion
    if (cfg.aon_clk_rst_vif.rst_n && !cfg.aon_intr_vif.pins) begin
      @(negedge cfg.aon_clk_rst_vif.rst_n or cfg.aon_intr_vif.pins);
    end
```

**Comment by ctopal** [comment]
Latest updates on the PR:

- A functioning scoreboard that counts according to the configuration added. It does the checking of interrupt interfaces (relaxed by 2 cycles because of synchronising between different clock domains) and calculates the remaining clock cycles.
- Restored `aon_core_if` to it's old version because it's not yet used.
- Some enhancements in the code (thanks Rupert for your suggestions!)

File: hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv [hunk 9]
```diff
@@ -10,7 +10,22 @@ class aon_timer_scoreboard extends cip_base_scoreboard #(
...
+
+  local bit  wdog_en;
+  local bit  wdog_num_update_due;
+  local uint wdog_count;
+  local uint bark_thold;
+  local uint bite_thold;
+
+  // expected values
+  local bit wkup_intr_status_exp = 0;
+  local bit wdog_intr_status_exp = 0;
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv:27]: I'd consider making this an array (`local bit intr_status_exp[2]`), which would make handling INTR_TEST and INTR_STATE a bit easier below. You could always define an enum with WKUP and WDOG and use it as an index?

File: hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv [hunk 10]
```diff
@@ -86,6 +130,107 @@ class aon_timer_scoreboard extends cip_base_scoreboard #(
...
+  // Task : compute_and_check_interrupt
+  // wait for expected # of clocks and check for interrupt state reg and pin
+  virtual task compute_and_check_interrupt();
+    uint wkup_num;
+    uint wdog_bark_num;
+    uint wdog_bite_num;
+
+    fork
+      begin
+        forever begin : compute_num_clks
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv:142]: Nit: I think I'd probably factor this out into a separate task and call it directly from `run_phase`. It will lose a level of `fork/join` and then the remaining "check_interrupt" code won't need a fork at all: it can just be a forever loop.

File: hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv [hunk 11]
```diff
@@ -86,6 +130,107 @@ class aon_timer_scoreboard extends cip_base_scoreboard #(
...
+        end // compute_num_clks
+      end
+    join_none
+
+    // TODO: Add coverage related changes
+    forever begin : wait_for_interrupt
+      @(wkup_en or wdog_en or under_reset);
+      wait(under_reset == 0);
+      // fork a thread for enabled timer
+      fork
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv:165]: Yikes! This fork/join stuff is quite complicated. I think I'd suggest a slight refactor, which should make it a lot easier to understand. You could factor out the wkup timer (lines 169-190) and wdog timer (lines 195-228) into their own functions and then remove the "under_reset" checks from them and just do it here.

The code here would then look something like this (untested!):
```systemverilog
    forever begin
      wait (!under_reset);

      fork : isolation_fork
        fork
          wait (under_reset);
          run_wkup_timer();
          run_wdog_timer();
        join_any

        // run_wkup_timer and run_wdog_timer never return so if we've got here then we've gone into
        // reset. Kill the two timer processes then go around and wait until we come out of reset
        // again.
        disable fork;
      join
    end
```

As an example, `run_wkup_timer()` would then be something like this:
```systemverilog
  task run_wkup_timer();
    forever begin
      wait (wkup_en);
      fork
        begin
          // trying to count how many cycles we need to count
          uint count = 0;
          cfg.aon_clk_rst_vif.wait_clks(2);
          forever begin
            @cfg.aon_clk_rst_vif.cb;
            // reset the cycle counter when we update the cycle count needed
            count = wkup_num_update_due ? 0 : (count + 1);
            if (count >= wkup_num) break;
          end
          `uvm_info(`gfn, $sformatf("WKUP Timer expired check for interrupts"), UVM_LOW)
          wkup_intr_status_exp = 1'b1;
          cfg.aon_clk_rst_vif.wait_clks(2);
          `DV_CHECK_CASE_EQ(wkup_intr_status_exp,
                            cfg.aon_intr_vif.sample_pin(.idx(1)))
        end
        begin
          wait (!wkup_en);
          `uvm_info(`gfn, $sformatf("WKUP Timer disabled, quit scoring"), UVM_HIGH)
        end
      join_any
      disable fork;
    end
  endtask
```

File: hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv [hunk 12]
```diff
@@ -86,6 +130,107 @@ class aon_timer_scoreboard extends cip_base_scoreboard #(
...
+            begin
+              // trying to count how many cycles we need to count
+              uint count = 0;
+              cfg.aon_clk_rst_vif.wait_clks(2);
+              forever begin
+                @cfg.aon_clk_rst_vif.cb;
+                // reset the cycle counter when we update the cycle count needed
+                count = wkup_num_update_due ? 0 : (count + 1);
+                if (count >= wkup_num) break;
+              end
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv:177]: Would a while loop be easier?
```systemverilog
          while (count < wkup_num) begin
            @cfg.aon_clk_rst_vif.cb;
            // reset the cycle counter when we update the cycle count needed
            count = wkup_num_update_due ? 0 : (count + 1);
          end
```
This has a different behaviour if `wkup_num` is zero: which one do you want?

File: hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv [hunk 13]
```diff
@@ -86,6 +130,107 @@ class aon_timer_scoreboard extends cip_base_scoreboard #(
...
+      @(wkup_en or wdog_en or under_reset);
+      wait(under_reset == 0);
+      // fork a thread for enabled timer
+      fork
+        if (wkup_en) begin
+          fork
+            begin
+              // trying to count how many cycles we need to count
+              uint count = 0;
+              cfg.aon_clk_rst_vif.wait_clks(2);
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv:171]: Could you add a comment explaining why we have this 2-cycle wait?

File: hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv [hunk 14]
```diff
@@ -86,6 +130,107 @@ class aon_timer_scoreboard extends cip_base_scoreboard #(
...
+                wdog_intr_status_exp = 1'b1;
+              end
+              if (count >= wdog_bite_num) begin
+                wdog_rst_req_exp = 1'b1;
+              end
+              cfg.aon_clk_rst_vif.wait_clks(2);
+              `DV_CHECK_CASE_EQ(wdog_intr_status_exp,
+                              cfg.aon_intr_vif.sample_pin(.idx(1)))
+              `DV_CHECK_CASE_EQ(wdog_rst_req_exp,
+                              cfg.aon_intr_vif.sample_pin(.idx(0)))
```
Comment (rswarbrick) [inline hw/ip/aon_timer/dv/env/aon_timer_scoreboard.sv:215]: Nit: Indentation of 2nd and 4th line

File: hw/ip/aon_timer/rtl/aon_timer.sv [hunk 15]
```diff
@@ -167,7 +167,7 @@ module aon_timer import aon_timer_reg_pkg::*;
   assign wkup_req_o = reg2hw.wkup_cause.q;
 
   // The wakeup signal is not latched in the pwrmgr so must be held until acked by software
-  `ASSERT(WkupStable_A, wkup_req_o |=> wkup_req_o ||
+  `ASSERT(WkupStable_A, wkup_req_o |-> wkup_req_o ||
       $fell(reg2hw.wkup_cause.q) && !aon_sleep_mode, clk_aon_i, !rst_aon_ni)
```
Comment (rswarbrick) [inline hw/ip/aon_timer/rtl/aon_timer.sv:171]: This looks a bit odd to me. Can you explain why the change is needed? I think the previous version said something like "the request only drops if the cause drops or we're no longer in sleep mode". Is that false?

**Comment by ctopal** [comment]
Thanks @rswarbrick for the comments about scoreboard and sorry for the not so clean code! I tried to follow your suggestions and hopefully make the code more clear now. 

The change about the assertion turns out to be unnecessary since the errors were related with me forgetting to drive sleep mode signal after sending a wake up request (Although I still think we should at least update the description of the assertion to say basically: "Check if request being driven to low only happens when sleep mode is low".). Now I am driving the sleep mode input to low after seeing an interrupt, do you think it makes sense?

**Review by rswarbrick — APPROVED** [review]
Looks good to me!
